packages/stargate/src/stargateclient.ts

```diff
@@ -268,7 +279,61 @@ export class StargateClient {
     if (this.tmClient) this.tmClient.disconnect();
   }
 
-  public async broadcastTx(tx: Uint8Array): Promise<BroadcastTxResponse> {
+  public async broadcastTx(
+    tx: Uint8Array,
+    timeoutMs = 60_000,
+    pollIntervalMs = 3_000,
+  ): Promise<BroadcastTxResponse> {
+    const txId = toHex(sha256(tx));
+    let timedOut = false;
+    const txPollTimeout = setTimeout(() => {
+      timedOut = true;
+    }, timeoutMs);
+
+    const handlePrematureTimeout = async (): Promise<BroadcastTxResponse> => {
+      if (timedOut) {
+        throw new TimeoutError(
+          `Transaction with ID ${txId} was submitted but was not yet found on the chain. You might want to check later.`,
+          txId,
+        );
+      }
+      await sleep(pollIntervalMs);
+      const result = await this.getTx(txId);
+      return result
+        ? {
+            code: result.code,
+            height: result.height,
+            rawLog: result.rawLog,
+            transactionHash: txId,
+          }
+        : handlePrematureTimeout();
+    };
+
+    return new Promise((resolve, reject) =>
+      this.broadcastTxCommit(tx)
+        .catch((error) => {
+          try {
+            const errorJson = JSON.parse(error.message);
+            if (
+              // -32603 is "Internal Error"
+              // https://github.com/tendermint/tendermint/blob/v0.34.0/rpc/jsonrpc/types/types.go#L236
+              errorJson.code === -32603 &&
+              /timed out waiting for tx to be included in a block/i.test(errorJson.data)
```

Comment:
If this check ever becomes a problem we can switch to non-committing boradcasts (sync) and start the polling right away.

Reply:
Updated in line with https://github.com/cosmos/cosmjs/issues/739#issuecomment-818587639